File: tools/mm2_codec.py
#!/usr/bin/env python3
"""MM2 codec tool for items.dat and str.dat.

Usage:
  python tools/mm2_codec.py items-decode <items.dat> <items.csv>
  python tools/mm2_codec.py items-encode <items.csv> <items.dat>
  python tools/mm2_codec.py str-decode <str.dat> <text.txt>
  python tools/mm2_codec.py str-encode <text.txt> <str.dat>
"""

from __future__ import annotations

import csv
import struct
import sys
from pathlib import Path

ITEM_RECORD_SIZE = 20
ITEM_NAME_SIZE = 12

ITEM_HEADER = [
    "index",
    "name",
    "separator",
    "forbidden_classes",
    "bonus_type",
    "bonus_amount",
    "effect_type",
    "effect_amount",
    "damage",
    "pad",
    "gold",
]


def _usage() -> int:
    print(__doc__.strip())
    return 1


def items_decode(in_path: Path, out_path: Path) -> None:
    data = in_path.read_bytes()
    if len(data) % ITEM_RECORD_SIZE != 0:
        raise ValueError(
            f"items size {len(data)} is not a multiple of {ITEM_RECORD_SIZE}"
        )

    rows = []
    record_count = len(data) // ITEM_RECORD_SIZE
    for idx in range(record_count):
        off = idx * ITEM_RECORD_SIZE
        rec = data[off : off + ITEM_RECORD_SIZE]
        name = rec[:ITEM_NAME_SIZE].decode("ascii", errors="strict").rstrip(" ")
        sep = rec[12]
        forbidden_classes = rec[13]  # set bit = class CANNOT use
        bonus = rec[14]
        effect = rec[15]
        damage = rec[16]
        pad = rec[17]
        gold = struct.unpack("<H", rec[18:20])[0]
        rows.append(
            {
                "index": idx,
                "name": name,
                "separator": sep,
                "forbidden_classes": forbidden_classes,
                "bonus_type": (bonus >> 4) & 0x0F,
                "bonus_amount": bonus & 0x0F,
                "effect_type": (effect >> 4) & 0x0F,
                "effect_amount": effect & 0x0F,
                "damage": damage,
                "pad": pad,
                "gold": gold,
            }
        )

    with out_path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=ITEM_HEADER)
        writer.writeheader()
        writer.writerows(rows)


def _u8(row: dict[str, str], key: str) -> int:
    v = int(row[key], 0)
    if not (0 <= v <= 0xFF):
        raise ValueError(f"{key} out of u8 range: {v}")
    return v


def _u16(row: dict[str, str], key: str) -> int:
    v = int(row[key], 0)
    if not (0 <= v <= 0xFFFF):
        raise ValueError(f"{key} out of u16 range: {v}")
    return v


def items_encode(in_path: Path, out_path: Path) -> None:
    out = bytearray()
    with in_path.open("r", newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        missing = [k for k in ITEM_HEADER if k not in (reader.fieldnames or [])]
        if missing:
            raise ValueError(f"missing CSV columns: {', '.join(missing)}")

        expected_index = 0
        for row in reader:
            idx = int(row["index"], 0)
            if idx != expected_index:
                raise ValueError(
                    f"row index mismatch: expected {expected_index}, got {idx}"
                )
            expected_index += 1

            name = row["name"]
            name_bytes = name.encode("ascii", errors="strict")
            if len(name_bytes) > ITEM_NAME_SIZE:
                raise ValueError(f"name too long ({len(name_bytes)}): {name!r}")
            name_field = name_bytes.ljust(ITEM_NAME_SIZE, b" ")

            sep = _u8(row, "separator")
            forbidden_classes = _u8(row, "forbidden_classes")
            bonus_type = _u8(row, "bonus_type")
            bonus_amount = _u8(row, "bonus_amount")
            effect_type = _u8(row, "effect_type")
            effect_amount = _u8(row, "effect_amount")
            if bonus_type > 0x0F or bonus_amount > 0x0F:
                raise ValueError("bonus_type/bonus_amount must be 0..15")
            if effect_type > 0x0F or effect_amount > 0x0F:
                raise ValueError("effect_type/effect_amount must be 0..15")

            damage = _u8(row, "damage")
            pad = _u8(row, "pad")
            gold = _u16(row, "gold")

            bonus = (bonus_type << 4) | bonus_amount
            effect = (effect_type << 4) | effect_amount

            rec = bytearray()
            rec.extend(name_field)
            rec.append(sep)
            rec.append(forbidden_classes)
            rec.append(bonus)
            rec.append(effect)
            rec.append(damage)
            rec.append(pad)
            rec.extend(struct.pack("<H", gold))
            if len(rec) != ITEM_RECORD_SIZE:
                raise AssertionError("bad record size")
            out.extend(rec)

    out_path.write_bytes(out)


def str_decode(in_path: Path, out_path: Path) -> None:
    src = in_path.read_bytes()
    decoded = bytearray()
    for b in src:
        if b == 0x01:
            decoded.append(0x0A)
        else:
            decoded.append((b + 0x1C) & 0xFF)
    out_path.write_bytes(decoded)


def str_encode(in_path: Path, out_path: Path) -> None:
    src = in_path.read_bytes()
    encoded = bytearray()
    for b in src:
        if b == 0x0A:
            encoded.append(0x01)
        elif b == 0x0D:
            # Ignore CR from CRLF input; LF is mapped above.
            continue
        else:
            encoded.append((b - 0x1C) & 0xFF)
    out_path.write_bytes(encoded)


def main(argv: list[str]) -> int:
    if len(argv) != 4:
        return _usage()

    cmd, src_s, dst_s = argv[1], argv[2], argv[3]
    src = Path(src_s)
    dst = Path(dst_s)

    if cmd == "items-decode":
        items_decode(src, dst)
    elif cmd == "items-encode":
        items_encode(src, dst)
    elif cmd == "str-decode":
        str_decode(src, dst)
    elif cmd == "str-encode":
        str_encode(src, dst)
    else:
        return _usage()
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: tools/mm2_codec.c
#include <errno.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ITEM_RECORD_SIZE 20
#define ITEM_NAME_SIZE 12

static void usage(void) {
    fprintf(stderr,
            "MM2 codec tool for items.dat and str.dat\n"
            "Usage:\n"
            "  mm2_codec items-decode <items.dat> <items.csv>\n"
            "  mm2_codec items-encode <items.csv> <items.dat>\n"
            "  mm2_codec str-decode <str.dat> <text.txt>\n"
            "  mm2_codec str-encode <text.txt> <str.dat>\n");
}

static long file_size(FILE *fp) {
    long cur = ftell(fp);
    if (cur < 0) {
        return -1;
    }
    if (fseek(fp, 0, SEEK_END) != 0) {
        return -1;
    }
    long size = ftell(fp);
    if (size < 0) {
        return -1;
    }
    if (fseek(fp, cur, SEEK_SET) != 0) {
        return -1;
    }
    return size;
}

static int parse_u8(const char *s, int *out) {
    char *end = NULL;
    errno = 0;
    long v = strtol(s, &end, 0);
    if (errno != 0 || end == s || *end != '\0' || v < 0 || v > 255) {
        return -1;
    }
    *out = (int)v;
    return 0;
}

static int parse_u16(const char *s, int *out) {
    char *end = NULL;
    errno = 0;
    long v = strtol(s, &end, 0);
    if (errno != 0 || end == s || *end != '\0' || v < 0 || v > 65535) {
        return -1;
    }
    *out = (int)v;
    return 0;
}

static int cmd_items_decode(const char *src_path, const char *dst_path) {
    FILE *in = fopen(src_path, "rb");
    FILE *out = NULL;
    if (!in) {
        fprintf(stderr, "open failed: %s\n", src_path);
        return 1;
    }
    out = fopen(dst_path, "wb");
    if (!out) {
        fprintf(stderr, "open failed: %s\n", dst_path);
        fclose(in);
        return 1;
    }

    long size = file_size(in);
    if (size < 0 || (size % ITEM_RECORD_SIZE) != 0) {
        fprintf(stderr, "invalid items.dat size: %ld\n", size);
        fclose(in);
        fclose(out);
        return 1;
    }

    fprintf(out,
            "index,name,separator,forbidden_classes,bonus_type,bonus_amount,effect_type,effect_amount,damage,pad,gold\n");

    uint8_t rec[ITEM_RECORD_SIZE];
    int index = 0;
    while (fread(rec, 1, ITEM_RECORD_SIZE, in) == ITEM_RECORD_SIZE) {
        char name[ITEM_NAME_SIZE + 1];
        for (int i = 0; i < ITEM_NAME_SIZE; i++) {
            name[i] = (char)rec[i];
        }
        name[ITEM_NAME_SIZE] = '\0';
        for (int i = ITEM_NAME_SIZE - 1; i >= 0; i--) {
            if (name[i] == ' ') {
                name[i] = '\0';
            } else {
                break;
            }
        }

        int separator = rec[12];
        int forbidden_classes = rec[13];  /* set bit = class CANNOT use */
        int bonus_type = (rec[14] >> 4) & 0x0F;
        int bonus_amount = rec[14] & 0x0F;
        int effect_type = (rec[15] >> 4) & 0x0F;
        int effect_amount = rec[15] & 0x0F;
        int damage = rec[16];
        int pad = rec[17];
        int gold = (int)rec[18] | ((int)rec[19] << 8);

        fprintf(out, "%d,%s,%d,%d,%d,%d,%d,%d,%d,%d,%d\n",
                index, name, separator, forbidden_classes, bonus_type, bonus_amount,
                effect_type, effect_amount, damage, pad, gold);
        index++;
    }

    fclose(in);
    fclose(out);
    return 0;
}

static int cmd_items_encode(const char *src_path, const char *dst_path) {
    FILE *in = fopen(src_path, "rb");
    FILE *out = NULL;
    if (!in) {
        fprintf(stderr, "open failed: %s\n", src_path);
        return 1;
    }
    out = fopen(dst_path, "wb");
    if (!out) {
        fprintf(stderr, "open failed: %s\n", dst_path);
        fclose(in);
        return 1;
    }

    char line[1024];
    if (!fgets(line, sizeof(line), in)) {
        fprintf(stderr, "empty CSV\n");
        fclose(in);
        fclose(out);
        return 1;
    }

    int expected_index = 0;
    while (fgets(line, sizeof(line), in)) {
        size_t n = strlen(line);
        while (n > 0 && (line[n - 1] == '\n' || line[n - 1] == '\r')) {
            line[--n] = '\0';
        }
        if (n == 0) {
            continue;
        }

        char *fields[11];
        int field_count = 0;
        char *tok = strtok(line, ",");
        while (tok && field_count < 11) {
            fields[field_count++] = tok;
            tok = strtok(NULL, ",");
        }
        if (field_count != 11) {
            fprintf(stderr, "bad CSV row, expected 11 columns\n");
            fclose(in);
            fclose(out);
            return 1;
        }

        int idx = 0;
        int separator = 0;
        int forbidden_classes = 0;
        int bonus_type = 0;
        int bonus_amount = 0;
        int effect_type = 0;
        int effect_amount = 0;
        int damage = 0;
        int pad = 0;
        int gold = 0;

        if (parse_u16(fields[0], &idx) != 0 || idx != expected_index) {
            fprintf(stderr, "index mismatch at row %d\n", expected_index);
            fclose(in);
            fclose(out);
            return 1;
        }
        expected_index++;

        const char *name = fields[1];
        size_t name_len = strlen(name);
        if (name_len > ITEM_NAME_SIZE) {
            fprintf(stderr, "name too long: %s\n", name);
            fclose(in);
            fclose(out);
            return 1;
        }

        if (parse_u8(fields[2], &separator) != 0 ||
            parse_u8(fields[3], &forbidden_classes) != 0 ||
            parse_u8(fields[4], &bonus_type) != 0 ||
            parse_u8(fields[5], &bonus_amount) != 0 ||
            parse_u8(fields[6], &effect_type) != 0 ||
            parse_u8(fields[7], &effect_amount) != 0 ||
            parse_u8(fields[8], &damage) != 0 ||
            parse_u8(fields[9], &pad) != 0 ||
            parse_u16(fields[10], &gold) != 0) {
            fprintf(stderr, "numeric parse error\n");
            fclose(in);
            fclose(out);
            return 1;
        }

        if (bonus_type > 0x0F || bonus_amount > 0x0F ||
            effect_type > 0x0F || effect_amount > 0x0F) {
            fprintf(stderr, "nibble fields must be 0..15\n");
            fclose(in);
            fclose(out);
            return 1;
        }

        uint8_t rec[ITEM_RECORD_SIZE];
        memset(rec, 0, sizeof(rec));
        memcpy(rec, name, name_len);
        for (size_t i = name_len; i < ITEM_NAME_SIZE; i++) {
            rec[i] = ' ';
        }
        rec[12] = (uint8_t)separator;
        rec[13] = (uint8_t)forbidden_classes;
        rec[14] = (uint8_t)((bonus_type << 4) | bonus_amount);
        rec[15] = (uint8_t)((effect_type << 4) | effect_amount);
        rec[16] = (uint8_t)damage;
        rec[17] = (uint8_t)pad;
        rec[18] = (uint8_t)(gold & 0xFF);
        rec[19] = (uint8_t)((gold >> 8) & 0xFF);

        if (fwrite(rec, 1, ITEM_RECORD_SIZE, out) != ITEM_RECORD_SIZE) {
            fprintf(stderr, "write failed\n");
            fclose(in);
            fclose(out);
            return 1;
        }
    }

    fclose(in);
    fclose(out);
    return 0;
}

static int cmd_str_decode(const char *src_path, const char *dst_path) {
    FILE *in = fopen(src_path, "rb");
    FILE *out = NULL;
    int ch;
    if (!in) {
        fprintf(stderr, "open failed: %s\n", src_path);
        return 1;
    }
    out = fopen(dst_path, "wb");
    if (!out) {
        fprintf(stderr, "open failed: %s\n", dst_path);
        fclose(in);
        return 1;
    }
    while ((ch = fgetc(in)) != EOF) {
        uint8_t b = (uint8_t)ch;
        uint8_t o = (b == 0x01) ? 0x0A : (uint8_t)((b + 0x1C) & 0xFF);
        if (fputc(o, out) == EOF) {
            fprintf(stderr, "write failed\n");
            fclose(in);
            fclose(out);
            return 1;
        }
    }
    fclose(in);
    fclose(out);
    return 0;
}

static int cmd_str_encode(const char *src_path, const char *dst_path) {
    FILE *in = fopen(src_path, "rb");
    FILE *out = NULL;
    int ch;
    if (!in) {
        fprintf(stderr, "open failed: %s\n", src_path);
        return 1;
    }
    out = fopen(dst_path, "wb");
    if (!out) {
        fprintf(stderr, "open failed: %s\n", dst_path);
        fclose(in);
        return 1;
    }
    while ((ch = fgetc(in)) != EOF) {
        uint8_t b = (uint8_t)ch;
        if (b == 0x0D) {
            continue;
        }
        uint8_t o = (b == 0x0A) ? 0x01 : (uint8_t)((b - 0x1C) & 0xFF);
        if (fputc(o, out) == EOF) {
            fprintf(stderr, "write failed\n");
            fclose(in);
            fclose(out);
            return 1;
        }
    }
    fclose(in);
    fclose(out);
    return 0;
}

int main(int argc, char **argv) {
    if (argc != 4) {
        usage();
        return 1;
    }

    const char *cmd = argv[1];
    const char *src = argv[2];
    const char *dst = argv[3];

    if (strcmp(cmd, "items-decode") == 0) {
        return cmd_items_decode(src, dst);
    }
    if (strcmp(cmd, "items-encode") == 0) {
        return cmd_items_encode(src, dst);
    }
    if (strcmp(cmd, "str-decode") == 0) {
        return cmd_str_decode(src, dst);
    }
    if (strcmp(cmd, "str-encode") == 0) {
        return cmd_str_encode(src, dst);
    }

    usage();
    return 1;
}
